# PYTHON BUILT-IN FUNCTIONS

### *A Comprehensive, Line-by-Line Reference Guide*

Every Built-in Function Explained: Syntax, Purpose, Code Examples and Best Practices

**Kajola Gbenga**  
*Data Scientist & AI Engineer*

---

## Table of Contents

- 1. Introduction to Python Built-in Functions
- 2. Setting Up the Python Environment
- 3. Type Conversion and Constructor Functions
- 4. Numeric and Math Functions
- 5. Iteration and Sequence Functions
- 6. String, Character and Representation Functions
- 7. Input/Output Functions
- 8. Object-Oriented and Class Functions
- 9. Attribute, Namespace and Scope Functions
- 10. Functional and Utility Helper Functions
- 11. Execution, Introspection and Debugging Functions
- 12. Full Worked Example (Combining Built-ins, End-to-End)
- 13. Best Practices When Using Built-in Functions
- 14. Common Pitfalls and How to Avoid Them
- 15. Quick-Reference Cheat Sheet (All 71 Built-ins A-Z)
- 16. Conclusion

# 1. Introduction to Python Built-in Functions

## 1.1 What Are Built-in Functions?

Python ships with a set of functions that are always available in every script, module or interactive session, without needing an import statement. These are the built-in functions, and as of modern Python (3.10+) there are 71 of them, ranging from everyday tools like print() and len() to lower-level utilities like memoryview() and __import__().

Built-in functions live in the builtins module, which Python loads automatically into every namespace. They are implemented in C in the reference CPython interpreter, which is part of why they are fast and reliable compared to writing equivalent logic by hand.

## 1.2 Why Mastering Built-ins Matters

- They save time: a single call to sorted() or sum() replaces many lines of manual looping code.
- They are optimized: built-ins are implemented in C and are almost always faster than a hand-written Python equivalent.
- They are the vocabulary of idiomatic Python: reading other people's code, libraries such as pandas and numpy, and interview questions all assume fluency with these functions.
- They reduce bugs: using isinstance() instead of comparing type() directly, or using enumerate() instead of manually tracking an index, avoids whole categories of mistakes.
- They form the foundation for everything else: functions like iter(), next(), __import__() and type() are the mechanisms that make loops, imports and classes work under the hood.

## 1.3 How This Guide Is Organized

All 71 built-in functions are grouped into nine functional categories: type conversion and constructors, numeric and math operations, iteration and sequences, string and character representation, input/output, object-oriented and class tools, attribute and scope inspection, functional/utility helpers, and execution/introspection.

Each function is presented with its signature, a plain-language explanation of what it does and when to reach for it, and a short runnable example. The most commonly misused or conceptually dense functions additionally receive a line-by-line breakdown table.

> **How to use this guide:** How to use this guide: read a category top to bottom the first time through, then use Section 15 (Quick-Reference Cheat Sheet) and the Table of Contents afterward to jump straight to any single function.

# 2. Setting Up the Python Environment

## 2.1 No Installation Required

Unlike pandas, numpy or seaborn, built-in functions require no pip install and no import statement. They are available the moment the Python interpreter starts, whether that is a plain .py script, a terminal REPL session, Jupyter Notebook, JupyterLab, VS Code or Google Colab.

The only requirement is a working Python 3 interpreter. This guide assumes Python 3.10 or later, since aiter() and anext() were added in Python 3.10; every other function described here has existed since at least Python 3.0.

## 2.2 Checking Your Python Version

In [1]:
import sys
print(sys.version)

3.12.3 (main, Mar  3 2026, 12:15:18) [GCC 13.3.0]


| Code Line | Explanation |
|---|---|
| `import sys` | Loads the standard library module that exposes interpreter details such as the running Python version. |
| `print(sys.version)` | Displays the exact Python version string, useful for confirming aiter()/anext() support (3.10+). |

## 2.3 Finding Help From Inside Python

Python's own help() built-in (covered fully in Section 11) can describe any other built-in function on demand, for example help(sorted), making the interpreter itself a live reference alongside this guide.

# 3. Type Conversion and Constructor Functions

These 14 functions either convert a value from one type to another, or construct a new object of a given type from scratch. They are usually the very first built-ins a Python learner meets, because every dataset read from a file, a form, or an API arrives as text and must be converted to the right type before it is useful.

## Summary Table

| Function | Description |
|---|---|
| `int(x, base=10)` | Converts a number or numeric string to an integer; base lets you parse binary, octal or hex strings. |
| `float(x)` | Converts a number or numeric string to a floating-point number. |
| `complex(real, imag)` | Creates a complex number with a real and an imaginary part. |
| `bool(x)` | Converts any value to True or False using Python's truthiness rules. |
| `str(obj)` | Converts any object to its human-readable string representation. |
| `list(iterable)` | Builds a new, mutable list from any iterable. |
| `tuple(iterable)` | Builds a new, immutable tuple from any iterable. |
| `dict(**kwargs)` | Builds a new dictionary from key/value pairs, an iterable of pairs, or keyword arguments. |
| `set(iterable)` | Builds a new, mutable, unordered collection of unique items. |
| `frozenset(iterable)` | Builds an immutable, hashable version of a set. |
| `bytes(source)` | Creates an immutable sequence of bytes, used for binary data. |
| `bytearray(source)` | Creates a mutable sequence of bytes. |
| `memoryview(obj)` | Creates a memory-efficient view over a bytes-like object without copying it. |
| `object()` | Creates a new instance of the base class from which every other Python class inherits. |

## 3.1 Converting User Input Safely

In [2]:
raw_age = "27"
age = int(raw_age)
price_text = "19.99"
price = float(price_text)
is_adult = bool(age >= 18)
print(age, price, is_adult)

27 19.99 True


| Code Line | Explanation |
|---|---|
| `int(raw_age)` | Parses the string "27" into the integer 27; raises ValueError if the text is not a valid whole number. |
| `float(price_text)` | Parses "19.99" into the float 19.99, preserving the decimal portion that int() would discard. |
| `bool(age >= 18)` | Evaluates the comparison to True or False, then bool() simply confirms/returns that boolean value. |

## 3.2 Building Collections From an Iterable

In [3]:
letters = list("hello")
coords = tuple([10, 20])
unique_ids = set([1, 2, 2, 3, 3, 3])
frozen = frozenset(unique_ids)
print(letters, coords, unique_ids, frozen)

['h', 'e', 'l', 'l', 'o'] (10, 20) {1, 2, 3} frozenset({1, 2, 3})


| Code Line | Explanation |
|---|---|
| `list("hello")` | Iterates over the string character by character and collects each character into a new list: ['h','e','l','l','o']. |
| `tuple([10, 20])` | Wraps the list's elements into an immutable tuple (10, 20); the original list is unaffected. |
| `set([1, 2, 2, 3, 3, 3])` | Discards duplicate values automatically, leaving the unique elements {1, 2, 3} in no guaranteed order. |
| `frozenset(unique_ids)` | Produces a hashable, read-only copy of the set, which can then be used as a dictionary key or set element. |

## 3.3 dict() Constructor in Three Forms

In [4]:
d1 = dict(name="Ada", role="Engineer")
d2 = dict([("a", 1), ("b", 2)])
d3 = dict(d2, c=3)
print(d1, d2, d3)

{'name': 'Ada', 'role': 'Engineer'} {'a': 1, 'b': 2} {'a': 1, 'b': 2, 'c': 3}


| Code Line | Explanation |
|---|---|
| `dict(name=..., role=...)` | Builds a dictionary directly from keyword arguments: {'name': 'Ada', 'role': 'Engineer'}. |
| `dict([("a",1), ("b",2)])` | Builds a dictionary from an iterable of two-item (key, value) tuples. |
| `dict(d2, c=3)` | Copies all pairs from d2 and adds/overwrites with the keyword argument c=3, without mutating d2. |

## 3.4 Working With Binary Data

In [5]:
raw = bytes([72, 101, 108, 108, 111])
mutable_raw = bytearray(raw)
mutable_raw[0] = 74
view = memoryview(mutable_raw)
print(raw, mutable_raw, bytes(view))

b'Hello' bytearray(b'Jello') b'Jello'


| Code Line | Explanation |
|---|---|
| `bytes([72,101,108,108,111])` | Creates an immutable byte sequence from a list of integer byte values (ASCII for 'Hello'). |
| `bytearray(raw)` | Creates a mutable copy of those bytes that individual elements can be reassigned in place. |
| `mutable_raw[0] = 74` | Changes the first byte from 72 ('H') to 74 ('J'), which is only possible because bytearray is mutable. |
| `memoryview(mutable_raw)` | Creates a zero-copy view onto the same underlying buffer, useful for large binary data where copying would be wasteful. |

> **Note:** object() is rarely called directly in everyday code; its real importance is as the implicit base class of every Python class, which is why every custom object automatically has methods like __repr__ and __eq__ available.

# 4. Numeric and Math Functions

These seven functions perform common numeric operations that would otherwise require importing the math module or writing manual loops. They work on ints, floats, and in several cases on any iterable of numbers.

## Summary Table

| Function | Description |
|---|---|
| `abs(x)` | Returns the absolute (non-negative) value of a number. |
| `divmod(a, b)` | Returns a tuple of (a // b, a % b) in a single call. |
| `pow(base, exp, mod=None)` | Raises base to the power exp, optionally taking a modulus efficiently. |
| `round(number, ndigits=None)` | Rounds a number to the given number of decimal places using banker's rounding. |
| `sum(iterable, start=0)` | Adds every item in an iterable together, starting from an optional initial value. |
| `min(iterable_or_args)` | Returns the smallest item in an iterable or among several arguments. |
| `max(iterable_or_args)` | Returns the largest item in an iterable or among several arguments. |

## 4.1 Everyday Numeric Operations

In [6]:
print(abs(-15))
print(divmod(17, 5))
print(pow(2, 10))
print(round(3.14159, 2))

15
(3, 2)
1024
3.14


| Code Line | Explanation |
|---|---|
| `abs(-15)` | Strips the negative sign, returning 15. |
| `divmod(17, 5)` | Returns (3, 2): 17 divided by 5 gives quotient 3 and remainder 2, in one call instead of two. |
| `pow(2, 10)` | Computes 2 raised to the 10th power, 1024, equivalent to 2 ** 10 but usable as a function object. |
| `round(3.14159, 2)` | Rounds to two decimal places, giving 3.14. |

## 4.2 Aggregating a Dataset Without pandas

In [7]:
scores = [88, 92, 79, 95, 63]
total = sum(scores)
average = total / len(scores)
highest = max(scores)
lowest = min(scores)
print(total, average, highest, lowest)

417 83.4 95 63


| Code Line | Explanation |
|---|---|
| `sum(scores)` | Adds every score together to get a running total, 417. |
| `total / len(scores)` | Divides the total by the count of scores to compute the mean. |
| `max(scores) / min(scores)` | Scan the list once each to find the single largest and smallest values without manual comparisons. |

## 4.3 min() and max() With a Key Function

In [8]:
students = [("Ada", 92), ("Bo", 88), ("Chi", 97)]
top_student = max(students, key=lambda s: s[1])
print(top_student)

('Chi', 97)


| Code Line | Explanation |
|---|---|
| `key=lambda s: s[1]` | Tells max() to compare each tuple by its second element (the score) rather than the whole tuple, so names never get compared to each other. |
| `max(students, key=...)` | Returns the entire ("Chi", 97) tuple whose score is the largest, not just the largest score. |

> **Note:** pow(base, exp, mod) with three arguments performs modular exponentiation efficiently, a technique used heavily in cryptography, and is much faster than computing pow(base, exp) % mod for large numbers.

# 5. Iteration and Sequence Functions

This is the largest category and arguably the most important for writing idiomatic, 'Pythonic' code. These 14 functions let you loop, transform, filter, combine and inspect sequences without writing manual index-based for loops.

## Summary Table

| Function | Description |
|---|---|
| `len(obj)` | Returns the number of items in a sequence or collection. |
| `range(start, stop, step)` | Generates an immutable, memory-efficient sequence of integers. |
| `enumerate(iterable, start=0)` | Pairs each item in an iterable with its index. |
| `zip(*iterables)` | Combines multiple iterables element-wise into tuples. |
| `map(func, iterable)` | Applies a function to every item in an iterable, lazily. |
| `filter(func, iterable)` | Keeps only the items for which a function returns True, lazily. |
| `reversed(seq)` | Returns an iterator that yields a sequence's items in reverse order. |
| `sorted(iterable, key=None, reverse=False)` | Returns a new, sorted list from any iterable. |
| `iter(obj)` | Returns an iterator object from an iterable. |
| `next(iterator, default)` | Retrieves the next item from an iterator. |
| `all(iterable)` | Returns True if every item in an iterable is truthy (or the iterable is empty). |
| `any(iterable)` | Returns True if at least one item in an iterable is truthy. |
| `aiter(async_iterable)` | Returns an asynchronous iterator from an async iterable (Python 3.10+). |
| `anext(async_iterator, default)` | Retrieves the next item from an async iterator (Python 3.10+). |

## 5.1 enumerate() Instead of Manual Index Tracking

In [9]:
fruits = ["mango", "pawpaw", "banana"]
for index, fruit in enumerate(fruits, start=1):
    print(index, fruit)

1 mango
2 pawpaw
3 banana


| Code Line | Explanation |
|---|---|
| `enumerate(fruits, start=1)` | Produces pairs (1, 'mango'), (2, 'pawpaw'), (3, 'banana') so the loop never needs a manually incremented counter variable. |
| `for index, fruit in ...` | Unpacks each pair directly into two loop variables in a single line. |

## 5.2 zip() to Walk Two Lists Together

In [10]:
names = ["Ada", "Bo", "Chi"]
scores = [92, 88, 97]
combined = list(zip(names, scores))
print(combined)

[('Ada', 92), ('Bo', 88), ('Chi', 97)]


| Code Line | Explanation |
|---|---|
| `zip(names, scores)` | Pairs each name with the score at the same position, producing an iterator of tuples that stops as soon as the shorter input runs out. |
| `list(zip(...))` | Materializes the lazy zip iterator into a concrete list so it can be printed and reused. |

## 5.3 map() and filter() With Lambda

In [11]:
numbers = [1, 2, 3, 4, 5, 6]
squared = list(map(lambda n: n ** 2, numbers))
evens = list(filter(lambda n: n % 2 == 0, numbers))
print(squared, evens)

[1, 4, 9, 16, 25, 36] [2, 4, 6]


| Code Line | Explanation |
|---|---|
| `map(lambda n: n**2, numbers)` | Applies the squaring function to every element lazily, producing an iterator; list() forces it to evaluate. |
| `filter(lambda n: n % 2 == 0, numbers)` | Keeps only elements where the lambda returns True (even numbers), discarding the rest. |

## 5.4 sorted() With key and reverse

In [12]:
words = ["banana", "fig", "apple", "kiwi"]
by_length = sorted(words, key=len)
longest_first = sorted(words, key=len, reverse=True)
print(by_length, longest_first)

['fig', 'kiwi', 'apple', 'banana'] ['banana', 'apple', 'kiwi', 'fig']


| Code Line | Explanation |
|---|---|
| `sorted(words, key=len)` | Sorts by the value len() returns for each word rather than alphabetically, giving shortest word first. |
| `reverse=True` | Flips the sort order so the longest word comes first instead. |

## 5.5 iter() and next() Under the Hood

In [13]:
numbers = [10, 20, 30]
it = iter(numbers)
print(next(it))
print(next(it))
print(next(it, "no more items"))
print(next(it, "no more items"))

10
20
30
no more items


| Code Line | Explanation |
|---|---|
| `iter(numbers)` | Creates an iterator object from the list; the list itself is iterable but is not, by itself, an iterator. |
| `next(it)` | Pulls the next value out of the iterator each time it is called, remembering position between calls. |
| `next(it, "no more items")` | Supplies a default value instead of raising StopIteration once the iterator is exhausted. |

## 5.6 all() and any() for Validation

In [14]:
ages = [22, 34, 19, 41]
print(all(age >= 18 for age in ages))
print(any(age < 18 for age in ages))

True
False


| Code Line | Explanation |
|---|---|
| `all(age >= 18 for age in ages)` | Returns True only if every single age satisfies the condition; short-circuits on the first False. |
| `any(age < 18 for age in ages)` | Returns True if at least one age satisfies the condition; short-circuits on the first True. |

> **Note:** aiter() and anext() are the asynchronous counterparts of iter() and next(), used with 'async for' loops over objects that define __anext__; they matter mainly in asyncio-based code such as streaming API clients, not in typical synchronous data analysis scripts.

# 6. String, Character and Representation Functions

These eight functions convert between characters and their numeric codes, between numbers and different bases, and between objects and various textual representations. They are essential for text processing, encoding work, and debugging.

## Summary Table

| Function | Description |
|---|---|
| `repr(obj)` | Returns the official, unambiguous string representation of an object, ideally one that could recreate it. |
| `format(value, format_spec)` | Formats a single value according to a format specification string. |
| `ascii(obj)` | Like repr(), but escapes any non-ASCII characters. |
| `ord(char)` | Returns the Unicode code point (integer) of a single character. |
| `chr(code)` | Returns the character corresponding to a Unicode code point (the inverse of ord()). |
| `hex(n)` | Converts an integer to a lowercase hexadecimal string prefixed with 0x. |
| `oct(n)` | Converts an integer to an octal string prefixed with 0o. |
| `bin(n)` | Converts an integer to a binary string prefixed with 0b. |

## 6.1 str() vs repr()

In [15]:
import datetime
now = datetime.date(2026, 9, 1)
print(str(now))
print(repr(now))

2026-09-01
datetime.date(2026, 9, 1)


| Code Line | Explanation |
|---|---|
| `str(now)` | Gives the informal, human-readable form: '2026-09-01', the same text shown by print(). |
| `repr(now)` | Gives the developer-facing form: 'datetime.date(2026, 9, 1)', valid Python code that would recreate the object. |

## 6.2 ord() and chr() Are Inverses

In [16]:
code_point = ord("A")
character = chr(code_point + 1)
print(code_point, character)

65 B


| Code Line | Explanation |
|---|---|
| `ord("A")` | Returns 65, the Unicode code point for the uppercase letter A. |
| `chr(code_point + 1)` | Converts the number 66 back into a character, giving 'B'. |

## 6.3 Number Base Conversions

In [17]:
n = 255
print(bin(n))
print(oct(n))
print(hex(n))

0b11111111
0o377
0xff


| Code Line | Explanation |
|---|---|
| `bin(255)` | Returns '0b11111111', the binary representation of 255. |
| `oct(255)` | Returns '0o377', the octal representation of 255. |
| `hex(255)` | Returns '0xff', the hexadecimal representation of 255. |

## 6.4 format() for Precise Display

In [18]:
pi = 3.14159265
print(format(pi, ".2f"))
print(format(1234567, ","))
print(format(0.873, ".1%"))

3.14
1,234,567
87.3%


| Code Line | Explanation |
|---|---|
| `format(pi, ".2f")` | Formats the float to exactly two decimal places: '3.14'. |
| `format(1234567, ",")` | Inserts thousands separators: '1,234,567'. |
| `format(0.873, ".1%")` | Displays the fraction as a percentage with one decimal place: '87.3%'. |

> **Note:** str() itself is covered in Section 3 as a type-conversion constructor; this section focuses on the representation and encoding helpers that complement it.

# 7. Input/Output Functions

These three functions are how a Python program talks to the outside world: displaying output, reading keyboard input, and reading or writing files.

## Summary Table

| Function | Description |
|---|---|
| `print(*objects, sep=' ', end='\n')` | Writes text to standard output (the console). |
| `input(prompt='')` | Pauses execution, displays an optional prompt, and reads a line of text typed by the user. |
| `open(file, mode='r')` | Opens a file and returns a file object for reading, writing or appending. |

## 7.1 print() Beyond the Basics

In [19]:
name = "Gbenga"
score = 97
print("Name:", name, "Score:", score, sep=" | ", end="\n\n")
print("Next line")

Name: | Gbenga | Score: | 97

Next line


| Code Line | Explanation |
|---|---|
| `sep=" \| "` | Replaces the default single-space separator between printed items with ' \| '. |
| `end="\n\n"` | Replaces the default trailing newline with two newlines, adding an extra blank line after this print call. |

## 7.2 Reading and Validating User Input

In [20]:
# In an interactive session, this line pauses and waits for keyboard input:
# raw = input("Enter your age: ")
raw = "27"  # simulating what the user typed, so this cell runs unattended
if raw.isdigit():
    age = int(raw)
    print(f"You are {age} years old.")
else:
    print("Please enter digits only.")

You are 27 years old.


| Code Line | Explanation |
|---|---|
| `input("Enter your age: ")` | Displays the prompt, blocks execution, and returns whatever the user typed as a plain string, before Enter is pressed. |
| `raw.isdigit()` | Checks the string only contains digit characters before attempting a conversion, avoiding a crash from int(). |

## 7.3 Reading a File Safely With a Context Manager

In [21]:
with open("notes.txt", "w") as f:
    f.write("Line one\nLine two\n")

with open("notes.txt", "r") as f:
    contents = f.read()
print(contents)

Line one
Line two



| Code Line | Explanation |
|---|---|
| `open("notes.txt", "w")` | Opens (creating if needed) the file in write mode, which truncates any existing content. |
| `with ... as f:` | Uses a context manager so the file is automatically and safely closed even if an error occurs inside the block. |
| `open("notes.txt", "r")` | Reopens the same file in read mode, the default mode when none is specified. |
| `f.read()` | Reads the entire file content into a single string. |

> **Note:** Always prefer the 'with open(...) as f:' pattern over calling open() and close() manually; it guarantees the file handle is released even if the code between them raises an exception.

# 8. Object-Oriented and Class Functions

These seven functions support Python's object-oriented programming model: inspecting types, checking inheritance relationships, calling a parent class's methods, and defining special kinds of class members.

## Summary Table

| Function | Description |
|---|---|
| `type(obj)` | Returns the type (class) of an object; with three arguments, dynamically creates a new class. |
| `isinstance(obj, classinfo)` | Checks whether an object is an instance of a class or tuple of classes. |
| `issubclass(cls, classinfo)` | Checks whether a class is a subclass of another class or tuple of classes. |
| `super()` | Returns a proxy object that delegates method calls to a parent or sibling class. |
| `property(fget, fset, fdel)` | Creates a managed attribute with optional getter, setter and deleter functions. |
| `staticmethod(func)` | Marks a method as not receiving an implicit first argument (no self, no cls). |
| `classmethod(func)` | Marks a method as receiving the class itself (cls) as its first argument instead of an instance. |

## 8.1 type() and isinstance()

In [22]:
value = 42
print(type(value))
print(isinstance(value, int))
print(isinstance(value, (int, float)))

<class 'int'>
True
True


| Code Line | Explanation |
|---|---|
| `type(value)` | Returns the exact class of value, <class 'int'>. |
| `isinstance(value, int)` | Checks membership more flexibly than type(); it also returns True for subclasses of int. |
| `isinstance(value, (int, float))` | Accepts a tuple of types, returning True if value matches any one of them. |

## 8.2 super() to Extend a Parent Class

In [23]:
class Animal:
    def __init__(self, name):
        self.name = name
    def speak(self):
        return f"{self.name} makes a sound."

class Dog(Animal):
    def speak(self):
        base_message = super().speak()
        return base_message + " Specifically, it barks."

print(Dog("Rex").speak())

Rex makes a sound. Specifically, it barks.


| Code Line | Explanation |
|---|---|
| `class Dog(Animal):` | Declares Dog as a subclass that inherits Animal's __init__ and can override speak(). |
| `super().speak()` | Calls Animal's original speak() method rather than duplicating its logic, then Dog's version extends the result. |

## 8.3 property(), staticmethod() and classmethod() via Decorators

In [24]:
class Circle:
    count = 0
    def __init__(self, radius):
        self.radius = radius
        Circle.count += 1

    @property
    def area(self):
        return 3.14159 * self.radius ** 2

    @staticmethod
    def is_valid_radius(value):
        return value > 0

    @classmethod
    def unit_circle(cls):
        return cls(radius=1)

c = Circle(2)
print(c.area)
print(Circle.is_valid_radius(-3))
print(Circle.unit_circle().radius)

12.56636
False
1


| Code Line | Explanation |
|---|---|
| `@property` | Turns area into an attribute-like call: c.area instead of c.area(), computed fresh every access. |
| `@staticmethod` | Marks is_valid_radius as not needing self or cls; it behaves like a plain function namespaced inside the class. |
| `@classmethod` | Marks unit_circle as receiving the class (cls) instead of an instance, letting it construct and return a new Circle. |

> **Note:** In modern code, property(), staticmethod() and classmethod() are almost always used as decorators (@property, @staticmethod, @classmethod) rather than called directly as functions, though both forms are equivalent.

# 9. Attribute, Namespace and Scope Functions

These eight functions inspect or modify an object's attributes, or inspect the variables currently visible in a given scope. They are especially useful for debugging, writing generic/dynamic code, and building tools like serializers or ORMs.

## Summary Table

| Function | Description |
|---|---|
| `vars(obj=None)` | Returns an object's __dict__; with no argument, behaves like locals(). |
| `dir(obj=None)` | Returns an alphabetical list of an object's valid attributes and methods. |
| `hasattr(obj, name)` | Checks whether an object has a given attribute, without raising an error. |
| `getattr(obj, name, default)` | Retrieves an attribute by name (as a string), with an optional default. |
| `setattr(obj, name, value)` | Sets an attribute by name (as a string) to a given value. |
| `delattr(obj, name)` | Deletes an attribute by name (as a string). |
| `globals()` | Returns a dictionary representing the current global symbol table. |
| `locals()` | Returns a dictionary representing the current local symbol table. |

## 9.1 Dynamic Attribute Access

In [25]:
class Config:
    def __init__(self):
        self.debug = True
        self.timeout = 30

cfg = Config()
field = "timeout"
print(hasattr(cfg, field))
print(getattr(cfg, field))
setattr(cfg, field, 60)
print(getattr(cfg, field))
delattr(cfg, "debug")
print(hasattr(cfg, "debug"))

True
30
60
False


| Code Line | Explanation |
|---|---|
| `hasattr(cfg, field)` | Safely checks whether Config has an attribute named 'timeout' without risking an AttributeError. |
| `getattr(cfg, field)` | Reads the attribute whose name is stored in the string variable field, rather than writing cfg.timeout literally. |
| `setattr(cfg, field, 60)` | Updates that same attribute to 60, again by name rather than by literal cfg.timeout = 60. |
| `delattr(cfg, "debug")` | Removes the debug attribute from the instance entirely. |

## 9.2 vars() and dir() for Introspection

In [26]:
cfg2 = Config()
print(vars(cfg2))
print([name for name in dir(cfg2) if not name.startswith("__")])

{'debug': True, 'timeout': 30}
['debug', 'timeout']


| Code Line | Explanation |
|---|---|
| `vars(cfg2)` | Returns the object's instance attributes as a plain dictionary: {'debug': True, 'timeout': 30}. |
| `dir(cfg2)` | Lists every attribute and method name accessible on the object, including inherited ones; filtering out dunder names keeps the output readable. |

## 9.3 globals() and locals()

In [27]:
x = 10

def show_scopes():
    y = 20
    print("y" in locals())
    print("x" in globals())

show_scopes()

True
True


| Code Line | Explanation |
|---|---|
| `locals()` | Inside the function, returns a dictionary of only that function's local variables, here containing y. |
| `globals()` | Returns the module-level symbol table, here containing x, regardless of which function calls it. |

> **Note:** getattr(obj, name, default) is a safer alternative to a bare try/except AttributeError block when you simply want a fallback value for a possibly-missing attribute.

# 10. Functional and Utility Helper Functions

These four functions provide low-level utilities that support functional programming patterns, object identity checks, hashing, and slicing.

## Summary Table

| Function | Description |
|---|---|
| `callable(obj)` | Checks whether an object can be called like a function (i.e. it defines __call__). |
| `id(obj)` | Returns the unique integer identity of an object for its lifetime (its memory address in CPython). |
| `hash(obj)` | Returns the integer hash value of an object, used internally by dicts and sets. |
| `slice(start, stop, step)` | Creates a slice object that can be reused to index into sequences. |

## 10.1 callable() to Check Before Calling

In [28]:
def greet():
    return "hello"

values = [greet, 42, "text", len]
for v in values:
    print(v, callable(v))

<function greet at 0x7ff64958db20> True
42 False
text False
<built-in function len> True


| Code Line | Explanation |
|---|---|
| `callable(v)` | Returns True for functions like greet and len, and False for plain data values like 42 and 'text', without actually invoking anything. |

## 10.2 id() and Object Identity

In [29]:
a = [1, 2, 3]
b = a
c = [1, 2, 3]
print(id(a) == id(b))
print(a is b)
print(id(a) == id(c))
print(a is c)

True
True
False
False


| Code Line | Explanation |
|---|---|
| `id(a) == id(b)` | True: b was assigned directly from a, so both names refer to the exact same list object in memory. |
| `id(a) == id(c)` | False: c is a separate list that merely has equal contents; 'a == c' would be True but 'a is c' is False. |

## 10.3 Reusable slice() Objects

In [30]:
data = [10, 20, 30, 40, 50, 60]
first_three = slice(0, 3)
last_two = slice(-2, None)
print(data[first_three])
print(data[last_two])

[10, 20, 30]
[50, 60]


| Code Line | Explanation |
|---|---|
| `slice(0, 3)` | Creates a named slice object equivalent to writing data[0:3] directly, but reusable across multiple sequences. |
| `slice(-2, None)` | Creates a slice equivalent to data[-2:], selecting from the second-to-last element to the end. |

> **Note:** hash(obj) only works on immutable, hashable objects; calling hash() on a list raises TypeError, which is precisely why lists cannot be used as dictionary keys but tuples can.

# 11. Execution, Introspection and Debugging Functions

These six functions execute code dynamically, inspect the interpreter itself, or assist with debugging. They are powerful but should be used sparingly and carefully in production code, since exec() and eval() can introduce serious security risks if used on untrusted input.

## Summary Table

| Function | Description |
|---|---|
| `exec(code)` | Executes a string (or compiled code object) containing arbitrary Python statements. |
| `eval(expression)` | Evaluates a string containing a single Python expression and returns its result. |
| `compile(source, filename, mode)` | Compiles source code into a code or AST object that can be executed later with exec() or eval(). |
| `__import__(name)` | The low-level function that the import statement calls internally to load a module. |
| `help(obj)` | Launches the interactive help system, or prints an object's documentation. |
| `breakpoint()` | Drops into the interactive debugger (pdb by default) at the point where it is called. |

## 11.1 eval() for Simple Expressions

In [31]:
expression = "3 * (4 + 5)"
result = eval(expression)
print(result)

27


| Code Line | Explanation |
|---|---|
| `eval(expression)` | Parses and evaluates the string as a Python expression, returning the computed integer 27. |

## 11.2 exec() for Multi-Statement Code

In [32]:
code_block = """
total = 0
for i in range(5):
    total += i
print('Total:', total)
"""
exec(code_block)

Total: 10


| Code Line | Explanation |
|---|---|
| `exec(code_block)` | Runs the entire multi-line string as if it were regular Python source code, including the loop and the print() call inside it. |

## 11.3 help() From Code

In [33]:
help(str.strip)

Help on method_descriptor:

strip(self, chars=None, /) unbound builtins.str method
    Return a copy of the string with leading and trailing whitespace removed.

    If chars is given and not None, remove characters in chars instead.



| Code Line | Explanation |
|---|---|
| `help(str.strip)` | Prints str.strip's docstring directly to the console, exactly as typing help(str.strip) in an interactive session would. |

> **Note:** Both exec() and eval() will run any code they are given with the full privileges of the calling program. Never pass them user-supplied text (from a web form, an API request, etc.) without rigorous sandboxing; safer alternatives such as ast.literal_eval() exist for the narrow case of parsing literal data.

# 12. Full Worked Example (Combining Built-ins, End-to-End)

The following self-contained script analyzes a small roster of employees using nothing but built-in functions, no external libraries, to show how naturally they combine in real code.

In [34]:
employees = [
    {"name": "Ada", "dept": "Engineering", "salary": 250000},
    {"name": "Bo", "dept": "Sales", "salary": 180000},
    {"name": "Chi", "dept": "Engineering", "salary": 275000},
    {"name": "Dumi", "dept": "Marketing", "salary": 160000},
    {"name": "Efe", "dept": "Sales", "salary": 195000},
]

# 1. Basic counts and totals
print("Total employees:", len(employees))
salaries = [e["salary"] for e in employees]
print("Total payroll:", sum(salaries))
print("Highest paid:", max(employees, key=lambda e: e["salary"])["name"])
print("Lowest paid:", min(employees, key=lambda e: e["salary"])["name"])

# 2. Filtering and mapping
engineers = list(filter(lambda e: e["dept"] == "Engineering", employees))
names_upper = list(map(lambda e: e["name"].upper(), engineers))
print("Engineering team:", names_upper)

# 3. Sorting
by_salary_desc = sorted(employees, key=lambda e: e["salary"], reverse=True)
for rank, person in enumerate(by_salary_desc, start=1):
    print(rank, person["name"], person["salary"])

# 4. Validation with all() / any()
all_paid_fairly = all(e["salary"] >= 150000 for e in employees)
any_over_250k = any(e["salary"] > 250000 for e in employees)
print("Everyone above minimum:", all_paid_fairly)
print("Anyone over 250k:", any_over_250k)

# 5. Departments as a set, then formatted output
departments = sorted(set(e["dept"] for e in employees))
print("Departments:", departments)

for dept in departments:
    dept_total = sum(e["salary"] for e in employees if e["dept"] == dept)
    print(f"{dept}: {format(dept_total, ',')}")


Total employees:

 5
Total payroll: 1060000
Highest paid: Chi
Lowest paid: Dumi
Engineering team: ['ADA', 'CHI']
1 Chi 275000
2 Ada 250000
3 Efe 195000
4 Bo 180000
5 Dumi 160000
Everyone above minimum: True
Anyone over 250k: True
Departments: ['Engineering', 'Marketing', 'Sales']
Engineering: 525,000
Marketing: 160,000
Sales: 375,000


## Line-by-Line Breakdown

| Code Line | Explanation |
|---|---|
| `len(employees)` | Counts the number of dictionaries in the list, 5. |
| `sum(salaries)` | Adds every salary in the derived salaries list to compute total payroll. |
| `max(..., key=lambda e: e['salary'])` | Finds the whole employee dictionary with the highest salary, using the salary field as the comparison key. |
| `filter(lambda e: ..., employees)` | Keeps only employees whose dept equals 'Engineering', discarding everyone else. |
| `map(lambda e: e['name'].upper(), engineers)` | Transforms each remaining employee dictionary into just their uppercased name. |
| `sorted(employees, key=..., reverse=True)` | Produces a new list ordered from highest to lowest salary, leaving the original list unchanged. |
| `enumerate(by_salary_desc, start=1)` | Attaches a 1-based rank to each employee while iterating over the sorted list. |
| `all(...) / any(...)` | Run a condition across every employee, short-circuiting as soon as the answer is determined. |
| `sorted(set(e['dept'] for e in employees))` | Deduplicates department names with set(), then sorted() puts them in alphabetical order. |
| `format(dept_total, ',')` | Displays each department's payroll total with thousands separators for readability. |

# 13. Best Practices When Using Built-in Functions

- Prefer a built-in over a hand-written loop whenever one exists; sum(), max(), min(), sorted(), any() and all() are implemented in C and are both faster and more readable than the equivalent manual loop.
- Use isinstance() rather than comparing type(obj) == SomeClass directly, since isinstance() also respects inheritance and works correctly with subclasses.
- Prefer enumerate() over range(len(sequence)) whenever you need both the index and the item; it removes an entire class of off-by-one bugs.
- Use the three-argument getattr(obj, name, default) instead of a try/except AttributeError block for simple fallback lookups; it is shorter and communicates intent more clearly.
- Convert map() and filter() results to a list only when you actually need a list; if the result will only be iterated once, leaving them as lazy iterators saves memory on large datasets.
- Avoid exec() and eval() on any input that could originate from a user or the network; prefer ast.literal_eval() for parsing simple literal values safely.
- Always close files, or better, always open files using a 'with' block so open() resources are released automatically even when an exception occurs.
- When comparing two variables to see whether they are the exact same object (not just equal in value), use 'is' (which relies on id()) rather than '=='.

# 14. Common Pitfalls and How to Avoid Them

| Pitfall | Why It Happens / How to Avoid It |
|---|---|
| `Using round() and expecting classic rounding` | round() uses banker's rounding (round half to even), so round(2.5) is 2, not 3; use the decimal module if exact decimal rounding rules matter. |
| `Comparing floats to integers with round()` | round(3.14159, 0) still returns a float (3.0), not an int; wrap it in int() explicitly if a true integer is required. |
| `Assuming map()/filter() return a list` | Both return lazy iterator objects in Python 3; forgetting to wrap them in list() means printing them shows a generator/iterator repr instead of the values. |
| `Calling next() on an exhausted iterator without a default` | Without a second argument, next() raises StopIteration once an iterator is exhausted, which can crash a script if not caught or given a default. |
| `Using min()/max() on an empty sequence` | Both raise ValueError on an empty iterable unless a default= keyword argument is supplied (Python 3.4+). |
| `Mutable default arguments confused with object()/dict()` | dict() and list() called as default arguments are evaluated once at function definition time, not per call, a classic source of shared-state bugs. |
| `Forgetting eval()/exec() security risk` | Running eval() or exec() on untrusted text can execute arbitrary code with the same permissions as the running program. |
| `Using id() to compare values` | id() reflects identity, not equality; two unrelated objects with identical content will normally have different ids, so use '==' to compare values. |

# 15. Quick-Reference Cheat Sheet (All 71 Built-ins A-Z)

Every built-in function covered in this guide, listed alphabetically with a one-line reminder of what it does.

| Function | One-Line Description |
|---|---|
| `abs(x)` | Absolute value. |
| `aiter(x)` | Async iterator from an async iterable. |
| `all(it)` | True if every item is truthy. |
| `anext(x)` | Next item from an async iterator. |
| `any(it)` | True if any item is truthy. |
| `ascii(obj)` | repr() with non-ASCII characters escaped. |
| `bin(n)` | Binary string of an integer. |
| `bool(x)` | Convert to True/False. |
| `breakpoint()` | Enter the interactive debugger. |
| `bytearray(x)` | Mutable sequence of bytes. |
| `bytes(x)` | Immutable sequence of bytes. |
| `callable(obj)` | Check whether an object can be called. |
| `chr(n)` | Character for a Unicode code point. |
| `classmethod(f)` | Make a method receive the class (cls), not an instance. |
| `compile(src, fn, mode)` | Compile source into a code object. |
| `complex(re, im)` | Create a complex number. |
| `delattr(obj, name)` | Delete a named attribute. |
| `dict(...)` | Build a dictionary. |
| `dir(obj)` | List an object's attributes. |
| `divmod(a, b)` | Quotient and remainder as a tuple. |
| `enumerate(it)` | Pair items with their index. |
| `eval(expr)` | Evaluate a Python expression string. |
| `exec(code)` | Execute Python statements from a string. |
| `filter(f, it)` | Keep items where f(item) is True. |
| `float(x)` | Convert to a floating-point number. |
| `format(val, spec)` | Format a value with a format spec. |
| `frozenset(it)` | Build an immutable set. |
| `getattr(obj, name)` | Get a named attribute. |
| `globals()` | Current global symbol table. |
| `hasattr(obj, name)` | Check whether an attribute exists. |
| `hash(obj)` | Integer hash of an object. |
| `help(obj)` | Show documentation. |
| `hex(n)` | Hexadecimal string of an integer. |
| `id(obj)` | Unique identity of an object. |
| `input(prompt)` | Read a line of text from the user. |
| `int(x)` | Convert to an integer. |
| `isinstance(obj, cls)` | Check an object's type, inheritance-aware. |
| `issubclass(cls, base)` | Check a class inheritance relationship. |
| `iter(obj)` | Get an iterator from an iterable. |
| `len(obj)` | Number of items. |
| `list(it)` | Build a mutable list. |
| `locals()` | Current local symbol table. |
| `map(f, it)` | Apply f to every item, lazily. |
| `max(it)` | Largest item. |
| `memoryview(obj)` | Zero-copy view over a buffer. |
| `min(it)` | Smallest item. |
| `next(it)` | Next item from an iterator. |
| `object()` | Base class instance. |
| `oct(n)` | Octal string of an integer. |
| `open(file, mode)` | Open a file. |
| `ord(ch)` | Unicode code point of a character. |
| `pow(base, exp)` | Exponentiation, optional modulus. |
| `print(...)` | Write text to standard output. |
| `property(f)` | Create a managed attribute. |
| `range(start, stop)` | Sequence of integers. |
| `repr(obj)` | Unambiguous string representation. |
| `reversed(seq)` | Reverse iterator over a sequence. |
| `round(n, digits)` | Round to a number of decimal places. |
| `set(it)` | Build a mutable set of unique items. |
| `setattr(obj, name, val)` | Set a named attribute. |
| `slice(start, stop)` | Reusable slice object. |
| `sorted(it)` | New sorted list. |
| `staticmethod(f)` | Method with no implicit first argument. |
| `str(obj)` | Human-readable string representation. |
| `sum(it)` | Add every item together. |
| `super()` | Proxy to a parent class. |
| `tuple(it)` | Build an immutable tuple. |
| `type(obj)` | The class of an object. |
| `vars(obj)` | An object's __dict__. |
| `zip(*its)` | Combine iterables element-wise. |
| `__import__(name)` | Low-level module import mechanism. |

# 16. Conclusion

The 71 built-in functions covered in this guide are not an optional extra to learn after 'real' Python; they are the foundation the rest of the language and its entire ecosystem, including pandas, numpy and every third-party library, is built on top of. Fluency with them is what separates code that works from code that reads clearly and runs efficiently.

From simple everyday tools like len(), print() and range(), through data-shaping helpers like map(), filter(), zip() and sorted(), to the more advanced object and introspection tools like super(), getattr() and eval(), each function exists to remove a specific, recurring piece of boilerplate.

The strongest habit to take from this guide is to reach for a built-in first before writing a manual loop or a custom helper function. More often than not, Python already has a fast, well-tested, one-line answer waiting in the builtins namespace.

---

**Kajola Gbenga | Data Scientist & AI Engineer**